### ISSUE TITLE
`as` capture over a value/literal pattern binds to `str` instead of the literal type (regression in 0.0.53)

### ISSUE THREAD

**Issue by SebbyLaw** [astral-sh/ty#3852]
### Summary

A capture pattern (`as name`) wrapping a literal pattern binds the name to `str` instead of
the matched literal type, causing a false-positive `invalid-assignment`/`invalid-return-type`.

## Repro

https://play.ty.dev/8962b047-5369-49a4-9627-b3e67174365b

```python
from typing import Literal, reveal_type

type Size = Literal["32", "128", "384"]


def f(size: str | None) -> Size | None:
    match size:
        case "32" | "128" | "384" as s:
            reveal_type(s)  # 0.0.53: `str`   0.0.52: `@Todo`
            return s        # 0.0.53: error[invalid-return-type]: `str` not assignable to `Size | None`
        case _:
            return None
```

Expected: `s` narrows to `Literal["32", "128", "384"]` (assignable to `Size | None`); no error.
This worked in ty 0.0.52, and both mypy and pyright accept it:

- mypy: `reveal_type(s)` → `Literal['32'] | Literal['128'] | Literal['384']`, no errors
- pyright: type of `s` is `Literal['32', '128', '384']`, no errors

### Version

ty 0.0.53 (23f26f98a 2026-06-23)

**Comment by AlexWaygood** [comment]
The problem here is that the literal pattern dispatches via an equality comparison at runtime. That means that if you passed in an instance of a subclass of `str` (for example, a `StrEnum` variant) that compared equal to `"32"`, it would take the first branch here:

```py
from enum import StrEnum

class NotASize(StrEnum):
    THIRTY_TWO = "32"

f(NotASize.THIRTY_TWO)
```

but that `f()` call doesn't return an object that inhabits the type `Size | None` -- it returns an object of type `Literal[NotASize.THIRTY_TWO]`, which is disjoint from the type either `Size` or `None`.

This is ultimately the same as #1566
